File: go/judge.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

const gatewayURL = "https://ai-gateway.vercel.sh/v4/ai/evaluation-model"

func modelID() string {
	if id := trim(os.Getenv("JEV_MODEL")); id != "" {
		return id
	}
	return "typesafe-ai/jev"
}

func gatewayJudge(state any, questions map[string]any) (judgeResult, error) {
	key := trim(os.Getenv("AI_GATEWAY_API_KEY"))
	if key == "" {
		return judgeResult{}, errBad(500, "缺少环境变量 AI_GATEWAY_API_KEY")
	}
	timeout := positiveInt(os.Getenv("JEV_TIMEOUT_MS"), 20_000)
	retries := nonNegativeInt(os.Getenv("JEV_MAX_RETRIES"), 1)
	body, err := json.Marshal(map[string]any{
		"state":     state,
		"questions": questions,
		"providerOptions": map[string]any{
			"gateway": map[string]any{"zeroDataRetention": true},
		},
	})
	if err != nil {
		return judgeResult{}, errBad(500, "无法编码 Jev 请求")
	}

	var last error
	for attempt := 0; attempt <= retries; attempt++ {
		result, status, retry, callErr := postEvaluate(key, body, time.Duration(timeout)*time.Millisecond)
		if callErr == nil {
			return result, nil
		}
		last = callErr
		if !retry || attempt == retries {
			if status != 0 {
				return judgeResult{}, callErr
			}
			break
		}
	}
	if last == nil {
		last = errBad(502, "Jev 调用失败")
	}
	return judgeResult{}, last
}

func postEvaluate(key string, body []byte, timeout time.Duration) (judgeResult, int, bool, error) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, gatewayURL, bytes.NewReader(body))
	if err != nil {
		return judgeResult{}, 0, false, errBad(500, "无法创建 Jev 请求")
	}
	req.Header.Set("Authorization", "Bearer "+key)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("ai-model-id", modelID())
	req.Header.Set("ai-evaluation-model-specification-version", "4")
	req.Header.Set("ai-gateway-protocol-version", "0.0.1")
	req.Header.Set("ai-gateway-auth-method", "api-key")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		if ctx.Err() != nil {
			return judgeResult{}, 0, true, errBad(504, "Jev 调用超时（"+strconv.Itoa(int(timeout.Milliseconds()))+"ms）")
		}
		return judgeResult{}, 0, true, errBad(502, redact(err.Error()))
	}
	defer resp.Body.Close()
	payload, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode == 401 {
		return judgeResult{}, 401, false, errBad(401, "AI Gateway 拒绝了这个 key。请检查 AI_GATEWAY_API_KEY 是否仍有效。")
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		raw := strings.TrimSpace(string(payload))
		if strings.Contains(strings.ToLower(raw), "credit card") {
			return judgeResult{}, 403, false, errBad(403, "Gateway key 是有效的，但这个 Vercel 账号还没有绑定信用卡，AI Gateway 拒绝了调用。到 Vercel 的 AI 页面加上卡并解锁免费额度后再试。")
		}
		status := resp.StatusCode
		if status < 400 || status >= 600 {
			status = 502
		}
		retry := status == 408 || status == 429 || status >= 500
		if raw == "" {
			raw = "Jev 调用失败"
		}
		return judgeResult{}, status, retry, errBad(status, redact(raw))
	}
	var decoded map[string]any
	if err := json.Unmarshal(payload, &decoded); err != nil {
		return judgeResult{}, 502, false, errBad(502, "Jev 返回不是合法 JSON")
	}
	answers, _ := decoded["answers"].(map[string]any)
	if answers == nil {
		return judgeResult{}, 502, false, errBad(502, "Jev 返回里没有 answers")
	}
	meta, _ := decoded["providerMetadata"].(map[string]any)
	return judgeResult{
		Answers:          answers,
		ProviderMetadata: meta,
		ModelID:          modelID(),
		Usage:            readUsage(decoded["usage"]),
	}, 200, false, nil
}

func readUsage(v any) *TokenUsage {
	raw, _ := v.(map[string]any)
	if raw == nil {
		return nil
	}
	in := readCount(raw["inputTokens"])
	if in == nil {
		in = readCount(raw["promptTokens"])
	}
	out := readCount(raw["outputTokens"])
	if out == nil {
		out = readCount(raw["completionTokens"])
	}
	total := readCount(raw["totalTokens"])
	if total == nil && in != nil && out != nil {
		sum := *in + *out
		total = &sum
	}
	if in == nil && out == nil && total == nil {
		return nil
	}
	return &TokenUsage{InputTokens: in, OutputTokens: out, TotalTokens: total}
}

func readCount(v any) *float64 {
	n, ok := asFloat(v)
	if !ok || !finite(n) {
		return nil
	}
	return &n
}

func positiveInt(raw string, fallback int) int {
	n, err := strconv.Atoi(trim(raw))
	if err != nil || n <= 0 {
		return fallback
	}
	return n
}

func nonNegativeInt(raw string, fallback int) int {
	n, err := strconv.Atoi(trim(raw))
	if err != nil || n < 0 {
		return fallback
	}
	return n
}

File: rust/src/judge.rs
use crate::engine::JudgeResult;
use crate::error::{err_bad, redact, IntuitionError};
use serde_json::{Map, Value};
use std::sync::OnceLock;
use std::time::Duration;

const GATEWAY_URL: &str = "https://ai-gateway.vercel.sh/v4/ai/evaluation-model";

pub fn model_id() -> String {
    std::env::var("JEV_MODEL")
        .ok()
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| "typesafe-ai/jev".into())
}

pub fn gateway_judge(state: &Value, questions: &Value) -> Result<JudgeResult, IntuitionError> {
    let key = std::env::var("AI_GATEWAY_API_KEY").unwrap_or_default();
    let key = key.trim();
    if key.is_empty() {
        return Err(err_bad(500, "缺少环境变量 AI_GATEWAY_API_KEY"));
    }
    let timeout = positive_int("JEV_TIMEOUT_MS", 20_000);
    let retries = non_negative_int("JEV_MAX_RETRIES", 1);
    let body = serde_json::json!({
        "state": state,
        "questions": questions,
        "providerOptions": {"gateway": {"zeroDataRetention": true}},
    });
    let mut last = err_bad(502, "Jev 调用失败");
    for attempt in 0..=retries {
        match post_evaluate(key, &body, timeout) {
            Ok(result) => return Ok(result),
            Err((err, retry)) => {
                last = err;
                if !retry || attempt == retries {
                    break;
                }
            }
        }
    }
    Err(last)
}

fn post_evaluate(
    key: &str,
    body: &Value,
    timeout_ms: u64,
) -> Result<JudgeResult, (IntuitionError, bool)> {
    let response = client()
        .post(GATEWAY_URL)
        .timeout(Duration::from_millis(timeout_ms))
        .bearer_auth(key)
        .header("content-type", "application/json")
        .header("ai-model-id", model_id())
        .header("ai-evaluation-model-specification-version", "4")
        .header("ai-gateway-protocol-version", "0.0.1")
        .header("ai-gateway-auth-method", "api-key")
        .json(body)
        .send();
    let response = match response {
        Ok(response) => response,
        Err(err) if err.is_timeout() => {
            return Err((
                err_bad(504, format!("Jev 调用超时（{timeout_ms}ms）")),
                true,
            ))
        }
        Err(err) => return Err((err_bad(502, redact(&err.to_string())), true)),
    };
    let status = response.status().as_u16();
    let payload = response.text().unwrap_or_default();
    if status == 401 {
        return Err((
            err_bad(
                401,
                "AI Gateway 拒绝了这个 key。请检查 AI_GATEWAY_API_KEY 是否仍有效。",
            ),
            false,
        ));
    }
    if !(200..300).contains(&status) {
        if payload.to_ascii_lowercase().contains("credit card") {
            return Err((err_bad(403, "Gateway key 是有效的，但这个 Vercel 账号还没有绑定信用卡，AI Gateway 拒绝了调用。到 Vercel 的 AI 页面加上卡并解锁免费额度后再试。"), false));
        }
        let mapped = if (400..600).contains(&status) {
            status
        } else {
            502
        };
        let retry = status == 408 || status == 429 || status >= 500;
        let raw = if payload.trim().is_empty() {
            "Jev 调用失败".into()
        } else {
            redact(payload.trim())
        };
        return Err((err_bad(mapped, raw), retry));
    }
    let decoded: Value = serde_json::from_str(&payload)
        .map_err(|_| (err_bad(502, "Jev 返回不是合法 JSON"), false))?;
    let answers = decoded
        .get("answers")
        .and_then(|v| v.as_object())
        .cloned()
        .ok_or_else(|| (err_bad(502, "Jev 返回里没有 answers"), false))?;
    let provider_metadata = decoded
        .get("providerMetadata")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_else(Map::new);
    Ok(JudgeResult {
        answers,
        provider_metadata,
        model_id: model_id(),
        usage: read_usage(decoded.get("usage")),
    })
}

fn read_usage(v: Option<&Value>) -> Option<crate::engine::TokenUsage> {
    let raw = v?.as_object()?;
    let input = read_count(raw.get("inputTokens")).or_else(|| read_count(raw.get("promptTokens")));
    let output =
        read_count(raw.get("outputTokens")).or_else(|| read_count(raw.get("completionTokens")));
    let total = read_count(raw.get("totalTokens")).or_else(|| match (input, output) {
        (Some(a), Some(b)) => Some(a + b),
        _ => None,
    });
    if input.is_none() && output.is_none() && total.is_none() {
        None
    } else {
        Some(crate::engine::TokenUsage {
            input_tokens: input,
            output_tokens: output,
            total_tokens: total,
        })
    }
}

fn read_count(v: Option<&Value>) -> Option<f64> {
    v.and_then(crate::policy::as_f64).filter(|n| n.is_finite())
}

fn client() -> &'static reqwest::blocking::Client {
    static CLIENT: OnceLock<reqwest::blocking::Client> = OnceLock::new();
    CLIENT.get_or_init(|| {
        reqwest::blocking::Client::builder()
            .build()
            .expect("reqwest client")
    })
}

fn positive_int(name: &str, fallback: u64) -> u64 {
    std::env::var(name)
        .ok()
        .and_then(|s| s.trim().parse().ok())
        .filter(|n: &u64| *n > 0)
        .unwrap_or(fallback)
}

fn non_negative_int(name: &str, fallback: u64) -> u64 {
    std::env::var(name)
        .ok()
        .and_then(|s| s.trim().parse().ok())
        .unwrap_or(fallback)
}
